clang/test/AST/const-fpfeatures.cpp

```diff
@@ -79,3 +79,16 @@ float V7 = []() -> float {
   0x0.000001p0F);
 }();
 // CHECK: @V7 = {{.*}} float 1.000000e+00
+
+template<float V> struct L {
+  constexpr L() : value(V) {}
+  float value;
+};
+
+#pragma STDC FENV_ROUND FE_DOWNWARD
```

Comment:
I meant something like:

``` template<typename T, T V> void foo() {
#pragma STDC FENV_ROUND FE_DOWNWARD
T Val = V;
// Should be:  { float 0x3FB9999980000000 }
}
#pragma STDC FENV_ROUND FE_UPDWARD
foo<float, 0.1F>();

Reply:
Ah, I see that github ate the first line of my comment, see above, i'v eedited it.

Reply:
In this snippet `Val` in `foo<float, 0.1F>()` should be `float 0x3FB99999A0000000` (rounded upward), because the literal `0.1F` is in the region, where `#pragma STDC FENV_ROUND FE_UPDWARD` is in effect. The body of `foo` does not contain literals.

Reply:
> In this snippet `Val` in `foo<float, 0.1F>()` should be `float 0x3FB99999A0000000` (rounded upward), because the literal `0.1F` is in the region, where `#pragma STDC FENV_ROUND FE_UPDWARD` is in effect. The body of `foo` does not contain literals.

Hmm... ok, then the feature isn't quite what I expected it was.  Does the `FENV_ROUND` exclusively change the behavior of literals?  I thought it messed with math as well? (I thought it was on assignment, but perhaps not!).

I'm more concerned about a case where the point-of-instantiation and the template have different rounding modes.  But I see this patch is exclusively for literals, so we can let that go.

One thing to ensure: can we have an AST test (like your 'foo' example) that shows that they are different instantiations?  Are there any cases where we could get two instantiation-requests spelled differently, but because of different rounding modes, are the same instantiation?

Reply:
> Does the FENV_ROUND exclusively change the behavior of literals? I thought it messed with math as well?

Of course, FENV_ROUND affects any math, but this patch fixes misbehavior of literals only. And the note was about different expectation.

Assignment isn't an operation that depends on rounding mode, but changing the snippet to such:
```
template<typename T, T V> void foo() {
#pragma STDC FENV_ROUND FE_DOWNWARD
T Val = V + 1;
}
```
would make use of the rounding mode. In this case the addition in `foo` would be evaluated using rounding down, but the literal is converted to AST object using rounding up.

> I'm more concerned about a case where the point-of-instantiation and the template have different rounding modes.

Does this test address your concern: https://github.com/spavloff/llvm-project/blob/aeb6074444513587924106081213335f73ba6eb0/clang/test/AST/ast-dump-fpfeatures.cpp#L124-L143 ?

> One thing to ensure: can we have an AST test (like your 'foo' example) that shows that they are different instantiations? Are there any cases where we could get two instantiation-requests spelled differently, but because of different rounding modes, are the same instantiation?

All instantiations of a function template use the rounding mode specified in the template definition. Otherwise we would have bad things like two instantiations in different translation units that use different rounding modes. The rounding mode is attached to AST nodes in the AST representation for function templates, tests in `AST/ast-dump-fpfeatures.cpp` check this property.

Reply:
>Does this test address your concern: https://github.com/spavloff/llvm-project/blob/aeb6074444513587924106081213335f73ba6eb0/clang/test/AST/ast-dump-fpfeatures.cpp#L124-L143 ?

That one does not.  I'm more looking for:
```
template<typename T, T f> void foo(){}
#pragma STDC FENV_ROUND FE_DOWNWARD
foo<float, 0.1f>();
#pragma STDC FENV_ROUND FE_UPWARD
foo<float, 0.1f>();
```

Then confirming that there are two separate instantiations of `foo`.

>>All instantiations of a function template use the rounding mode specified in the template definition. Otherwise we would have bad things like two instantiations in different translation units that use different rounding modes. 

Right, that was very much my concern, the ODR implications of that are scary.  So jsut the 1 test above, and I think _I_ am good about this, though having @jcranmer-intel and/or @zahiraam review would also be appreciated.

Reply:
The interaction of these pragmas with C++ is underspecified, although for the most part, you can make pretty reasonable expectations about what the results should be. I'm also planning on bringing a paper to C++ to clarify the semantics somewhat, with the intent mostly being "just don't allow the pragma in most places."

That said, templates do add in all sorts of fun extra complexity, and there are at least three scenarios worth testing (with regards to interpretation of constants within the template body):
* Implicit template instantiation (you already test this)
* Explicit template instantiation
* Template specialization

I can definitely agree that implicit template instantiation should inherit from the rounding mode of the definition. Template specialization probably *shouldn't* inherit from the original template definition. Explicit instantiation... I'm not sure? I can argue that one either way.

Reply:
Explicit instantiation also should use the rounding mode at the point of definition. It is possible that the template is instantiated in two translation units and rounding mode is different at the points of instantiation.